src/Authentication/Authentication/TokenCache/TokenCacheStorage.cs

```diff
@@ -45,59 +53,73 @@ public static byte[] GetToken(string appId)
         /// </summary>
         /// <param name="appId">An app/client id.</param>
         /// <param name="accessToken">An access token.</param>
-        public static void SetToken(string appId, byte[] accessToken)
+        public static void SetToken(IAuthContext authContext, byte[] accessToken)
         {
-            if (string.IsNullOrEmpty(appId))
+            if (string.IsNullOrEmpty(authContext.ClientId))
             {
                 throw new ArgumentNullException(string.Format(
                     CultureInfo.CurrentCulture,
                     ErrorConstants.Message.NullOrEmptyParameter,
-                    nameof(appId)));
+                    nameof(authContext.ClientId)));
             }
             if (accessToken == null || accessToken.Length == 0)
             {
                 return ;
             }
 
-            if (Helpers.OperatingSystem.IsWindows())
+            if (authContext.ContextScope == ContextScope.Process)
             {
-                WindowsTokenCache.SetToken(appId, accessToken);
+                GraphSession.Instance.Token = accessToken;
             }
-            else if (Helpers.OperatingSystem.IsMacOS())
+            else if(authContext.ContextScope == ContextScope.CurrentUser)
             {
-                MacTokenCache.SetToken(appId, accessToken);
-            }
-            else
-            {
-                LinuxTokenCache.SetToken(appId, accessToken);
+                if (Helpers.OperatingSystem.IsWindows())
+                {
+                    WindowsTokenCache.SetToken(authContext.ClientId, accessToken);
+                }
+                else if (Helpers.OperatingSystem.IsMacOS())
+                {
+                    MacTokenCache.SetToken(authContext.ClientId, accessToken);
+                }
+                else
+                {
+                    LinuxTokenCache.SetToken(authContext.ClientId, accessToken);
+                }
```

Comment:
since this pattern seems to be repeating in at least 3 different places, do we have an interface where all three token caches implement, so that we can do something like:
```
var cache = Helpers.GetOSCache();
cache.SetToken(clientId, accessToken)
```

also for `DeleteToken` and `GetToken` respectively?

Reply:
This is a great suggestion, I'll add an interface for this.

Reply:
On problem though, C# 7 doesn't allow static interfaces.

Reply:
Singleton non-static classes can achieve the same thing I presume, but feel free to not take that change if that means too much refactoring for the scope of this fix.